source/bigint.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/bigint.h>
+
+#define BASE_BITS 32
+#define NIBBLES_PER_DIGIT ((BASE_BITS) / 4)
+#define LOWER_32_BIT_MASK 0xFFFFFFFF
+#define INT64_MIN_AS_HEX 0x8000000000000000
+
+/*
+ * Working set of invariants:
+ *
+ * (1) Negative zero is illegal
+ * (2) The only time leading (trailing) 0-value digits are allowed is a single instance to represent actual zero.
+ *
+ * Other proposals:
+ *
+ * (1) Functionally immutable API (no visible side-affects)
+ * (2) params may be temporarily altered during computation before being restored in order to prevent pointless copying
+ *   (most often around sign negation and operator switching)
+ *
+ * Dev ideas:
+ *
+ * (1) Internal arithmetic ops may use a "raw" interface to allow for subsequence operations used in multiply/divide.
+ */
+
+void aws_bigint_clean_up(struct aws_bigint *bigint) {
+    aws_array_list_clean_up(&bigint->digits);
+}
+
+static void s_advance_cursor_past_hex_prefix(struct aws_byte_cursor *hex_cursor) {
+    if (hex_cursor->len >= 2) {
+        const char *raw_ptr = (char *)hex_cursor->ptr;
+        if (raw_ptr[0] == '0' && (raw_ptr[1] == 'x' || raw_ptr[1] == 'X')) {
+            aws_byte_cursor_advance(hex_cursor, 2);
+        }
+    }
+}
+
+static void s_advance_cursor_to_non_zero(struct aws_byte_cursor *hex_cursor) {
+    while (hex_cursor->len > 0 && *hex_cursor->ptr == '0') {
```

Comment:
memchar?

Reply:
I don't know of a not-equal version, so don't see how that would help